### PR TITLE
workflows: integrate reference matcher in workflows

### PR DESCRIPTION
**Pull request by salmanmaq**
mappings: integrate reference matcher in Labs workflows

## Description
This integrates the reference matcher in the Labs workflow so that the reference matching is performed in Labs and not in Legacy.

## Related Issue
https://its.cern.ch/jira/browse/INSPIR-447

## Motivation and Context
The new reference matcher aims to improve citation counts over the Legacy reference matcher. It incorporates data citations as well.

## Checklist:
- [x] I have all the information that I need (if not, move to `RFC` and look for it).
- [ ] I linked the related issue(s) in the corresponding commit logs.
- [x] I wrote [good commit log messages](https://github.com/torvalds/subsurface-for-dirk/blob/5f15ad5a86ada3c5e574041a5f9d85235322dabb/README#L92-L119).
- [x] My code follows the code style of this project.
- [x] I've added any new docs if API/utils methods were added.
- [x] I have updated the existing documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: inspirehep/modules/workflows/tasks/refextract.py [hunk 1]
```diff
@@ -184,3 +189,60 @@ def extract_references_from_raw_ref(reference, custom_kbs_file=None):
...
+        reference: The reference metadata
+        config: The configurtaion(s) for InspireMatcher queries
+
+    Returns:
+        The record ID of the matched reference
+    """
+
+    config_default = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DEFAULT_CONFIG']
+    config_jcap_and_jhep = current_app.config['WORKFLOWS_REFERENCE_MATCHER_JHEP_AND_JCAP_CONFIG']
+    config_data = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DATA_CONFIG']
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/refextract.py]: Let's do one thing at a time: if we were to match on data with the current code then we will produce the wrong `$ref`. So we can't have this code in here yet, unless we solve that other issue first.

File: inspirehep/modules/workflows/tasks/refextract.py [hunk 2]
```diff
@@ -184,3 +189,60 @@ def extract_references_from_raw_ref(reference, custom_kbs_file=None):
...
+
+    journal_title = get_value(reference, 'reference.publication_info.journal_title')
+    if journal_title in ['JCAP', 'JHEP']:
+        try:
+            if get_value(reference, 'reference.publication_info.year'):
+                reference['reference']['publication_info']['year'] = str(
+                    reference['reference']['publication_info']['year'])
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/refextract.py]: Can you check if we really need this? These fields agree on the type, see https://github.com/inspirehep/inspire-next/blob/ffcd5904004a13090ea289d1380aaa2d03a056bd/inspirehep/modules/records/mappings/records/hep.json#L713-L716 and https://github.com/inspirehep/inspire-next/blob/ffcd5904004a13090ea289d1380aaa2d03a056bd/inspirehep/modules/records/mappings/records/hep.json#L893-L896.

File: inspirehep/modules/workflows/config.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                {
+                    'paths': [
+                        'reference.publication_info.journal_title',
+                        'reference.publication_info.journal_volume',
+                        'reference.publication_info.artid',
+                    ],
+                    'search_paths': [
+                        'publication_info.journal_title.raw',
+                        'publication_info.journal_volume',
+                        'publication_info.artid',
```
Comment (jacquerie) [inline inspirehep/modules/workflows/config.py]: Can you check if we still need these searches to be distinct, given the following commit by @iulianav? https://github.com/inspirehep/inspire-next/commit/09b1255ad0146a0d124a1a93dd1b1fba8ed6d16c#diff-472ccc26aabcd94f4c20bce673a6e90d

File: inspirehep/modules/workflows/config.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                        'reference.publication_info.journal_title',
+                        'reference.publication_info.journal_volume',
+                        'reference.publication_info.year',
+                        'reference.publication_info.artid',
+                    ],
+                    'search_paths': [
+                        'publication_info.journal_title.raw',
+                        'publication_info.journal_volume',
+                        'publication_info.year',
+                        'publication_info.artid',
```
Comment (jacquerie) [inline inspirehep/modules/workflows/config.py]: Same consideration as https://github.com/inspirehep/inspire-next/pull/3351/files#r182354877.

File: tests/integration/workflows/test_workflow_actions.py [hunk 5]
```diff
@@ -329,3 +400,53 @@ def test_normalize_journal_titles_unknown_journals_no_ref(workflow_app, insert_j
...
+    }
+
+    with workflow_app.app_context():
+        with mock.patch.dict(workflow_app.config, extra_config):
+            citing_doc_workflow_uuid = start('article', [citing_record])
+
+        citing_doc_eng = WorkflowEngine.from_uuid(citing_doc_workflow_uuid)
+        citing_doc_obj = citing_doc_eng.processed_objects[0]
+
+        assert citing_doc_obj.data['references'][7]['record'] == get_record_ref(insert_cited_record['control_number'], 'literature')
```
Comment (jacquerie) [inline tests/integration/workflows/test_workflow_actions.py]: I think that here we should follow the suggestion in https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html:
> Tests can avoid complexity by stating their inputs and outputs directly rather than computing them. Otherwise it's easy for tests to develop their own bugs.

File: inspirehep/modules/workflows/tasks/actions.py [hunk 6]
```diff
@@ -421,17 +422,19 @@ def refextract(obj, eng):
...
+        text_references = match_references(text_references)
 
     if len(pdf_references) == len(text_references) == 0:
         obj.log.info('No references extracted.')
-    elif len(pdf_references) >= len(text_references):
+    elif len(pdf_references) > len(text_references):
         obj.log.info('Extracted %d references from PDF.', len(pdf_references))
         obj.data['references'] = pdf_references
-    elif len(text_references) > len(pdf_references):
+    elif len(text_references) >= len(pdf_references):
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/actions.py]: Can you remind me why this changed behavior (preferring text references to PDF references when they are in the same amount) ?

File: inspirehep/modules/workflows/tasks/refextract.py [hunk 7]
```diff
@@ -184,3 +189,60 @@ def extract_references_from_raw_ref(reference, custom_kbs_file=None):
...
+        result = next(match(reference, config_data))
+        return result['_source']['control_number']
+    except StopIteration:
+        pass
+
+    try:
+        result = next(match(reference, config_default))
+        return result['_source']['control_number']
+    except StopIteration:
+        pass
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/refextract.py]: Repeating the same code pattern 3 times (especially this close to each other) should make you think that a refactoring is possible: https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming).

Currently this function is an example of this other "Rule-of-three": https://en.wikipedia.org/wiki/Planescape#Rule-of-Three : )

File: inspirehep/modules/workflows/tasks/refextract.py (hunk 1 above)
Comment (jacquerie) [reply inspirehep/modules/workflows/tasks/refextract.py]: To be clear: this applies to the current PR, but by the end of the sprint we **are** going to handle data references!

File: inspirehep/modules/workflows/config.py (hunk 3 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/config.py]: That should be the case. However, I do not see any keys named page_artid in Labs for the few records I have inspected. Am I missing something?

File: inspirehep/modules/workflows/tasks/actions.py (hunk 6 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/tasks/actions.py]: The assumption here is that the text references provided by the submitter potentially have a better quality than references extracted from PDF. So, if both of them manage to return the same amount of references, we prefer the text ones.

File: inspirehep/modules/workflows/tasks/refextract.py (hunk 1 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/tasks/refextract.py]: Do we handle the data issue as well in this PR, or should we try to merge this before doing that?

File: inspirehep/modules/workflows/tasks/refextract.py (hunk 7 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/tasks/refextract.py]: Will change.

File: inspirehep/modules/workflows/tasks/refextract.py (hunk 1 above)
Comment (jacquerie) [reply inspirehep/modules/workflows/tasks/refextract.py]: I'd say that we limit the scope of the current PR to Literature references, but in the very next PR we're going to handle Data references as well.

File: inspirehep/modules/workflows/config.py (hunk 3 above)
Comment (jacquerie) [reply inspirehep/modules/workflows/config.py]: This has to do with the thing I say often: "ES is not a JSON store". The `copy_to` directive **does not** enhance the JSON that you are sending as the source document, but instructs ES to index the field in another inverted index; so when you retrieve records from ES you get the source JSON back, which hasn't changed (in particular it doesn't contain `page_artid`), but searches against that field in the inverted index will work.

File: inspirehep/modules/workflows/tasks/actions.py (hunk 6 above)
Comment (jacquerie) [reply inspirehep/modules/workflows/tasks/actions.py]: Ok, makes sense. Nit: let's put this behavior change in a distinct commit, so if @michamos says that we did a mistake we can easily revert it (did we ever get his opinion on this?).

File: inspirehep/modules/workflows/config.py (hunk 3 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/config.py]: Ok, nice. I'll verify and make the necessary changes.

File: inspirehep/modules/workflows/tasks/actions.py (hunk 6 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/tasks/actions.py]: We didn't get his opinion on this. But that makes sense.

File: inspirehep/modules/workflows/config.py (hunk 3 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/config.py]: I checked this and the records with publication_info no longer match.

File: inspirehep/modules/workflows/tasks/refextract.py (hunk 2 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/tasks/refextract.py]: Verified to work without this.

File: inspirehep/modules/workflows/tasks/refextract.py (hunk 7 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/tasks/refextract.py]: Another comment on this, while we can do what you suggested, I think this would lead to a bit too much abstraction. What do you think about that?

Comment (jacquerie) [reply inspirehep/modules/workflows/tasks/refextract.py]: Mh, I see what you mean, but I would still do it. Let's see what @turtle321 thinks about this?

Comment (turtle321) [reply inspirehep/modules/workflows/tasks/refextract.py]: I woul probably try to avoid excetions for control flows if possible, thus I would have written:
```
result = next(match(reference, config_jcap_and_jhep), None)
if result:
    return result['_source']['control_number']
```

File: inspirehep/modules/workflows/config.py (hunk 3 above)
Comment (michamos) [reply inspirehep/modules/workflows/config.py]: didn't we want to search for both `artid` and `page_start` in `page_artid` instead of doing `artid -> artid`, `page_start -> page_start` only?

File: inspirehep/modules/workflows/config.py (hunk 4 above)
Comment (michamos) [reply inspirehep/modules/workflows/config.py]: Same also for my comment.

File: inspirehep/modules/workflows/tasks/actions.py [hunk 8]
```diff
@@ -400,19 +401,19 @@ def refextract(obj, eng):
...
         eng: a workflow engine.
 
     Returns:
         None
+
     """
     if 'references' in obj.data:
-        obj.log.info('Found references in metadata, extracting unextracted raw_refs')
-        obj.data['references'] = extract_references_from_raw_refs(obj.data['references'])
+        raw_references = extract_references_from_raw_refs(obj.data['references'])
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/actions.py]: why is this called `raw_references`? I would expect the result to be a list of parsed, structured references after processing by refextract. Maybe `extracted_raw_references`?

File: inspirehep/modules/workflows/tasks/actions.py (hunk 6 above)
Comment (michamos) [reply inspirehep/modules/workflows/tasks/actions.py]: Seems OK to me.

File: inspirehep/modules/workflows/tasks/refextract.py [hunk 9]
```diff
@@ -184,3 +189,46 @@ def extract_references_from_raw_ref(reference, custom_kbs_file=None):
...
+        references: Schema compliant list of references
+
+    Returns:
+        associated_references: The same list with the associated record ids (if any)
+    """
+
+    for i, reference in enumerate(references):
+        matched_recid = match_reference(reference)
+        if matched_recid:
+            references[i]['record'] = get_record_ref(matched_recid, 'literature')
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/refextract.py]: I would drop this function, and have a `match_reference` that does all the work for one reference, that is take a reference and spit out a reference, potentially with the `record` key set to the matched record.
You would then replace calls of `result = match_references(references)` by `result = [match_reference(ref) for ref in references]`.

File: inspirehep/modules/workflows/tasks/refextract.py [hunk 10]
```diff
@@ -184,3 +189,46 @@ def extract_references_from_raw_ref(reference, custom_kbs_file=None):
...
+
+    Returns:
+        The record ID of the matched reference
+    """
+
+    config_default = current_app.config['WORKFLOWS_REFERENCE_MATCHER_DEFAULT_CONFIG']
+    config_jcap_and_jhep = current_app.config['WORKFLOWS_REFERENCE_MATCHER_JHEP_AND_JCAP_CONFIG']
+
+    journal_title = get_value(reference, 'reference.publication_info.journal_title')
+    if journal_title in ['JCAP', 'JHEP']:
```
Comment (michamos) [inline inspirehep/modules/workflows/tasks/refextract.py]: for this to work properly, the `normalize_journal_titles` has to run before `refextract` in https://github.com/inspirehep/inspire-next/blob/7aad1074b4285b4e5dbf246c7d3cfd0f1ad166af/inspirehep/modules/workflows/workflows/article.py#L157-L158.

Comment (michamos) [inline inspirehep/modules/workflows/tasks/refextract.py]: it would be cleaner to have
```python
config = config_jcap_and_jhep if journal_title in ['JCAP', 'JHEP'] else config_default
result = next(match(reference, config_jcap_and_jhep), None)
if result:
    return result['_source']['control_number']
return None
```

File: tests/integration/workflows/helpers/calls.py [hunk 11]
```diff
@@ -167,6 +167,28 @@ def generate_record():
...
+            'oai_arxiv_citing_record.json'
+        )
+    ))
+
+    categories = {'core': [], 'non-core': []}
+    for eprint in json_data.get('arxiv_eprints', []):
+        categories['core'].extend(eprint.get('categories', []))
+
+    if 'preprint_date' in json_data:
+        json_data['preprint_date'] = datetime.date.today().isoformat()
```
Comment (michamos) [inline tests/integration/workflows/helpers/calls.py]: MINOR it would have been cleaner to use https://github.com/spulec/freezegun here instead, but it's OK for now.

File: tests/integration/workflows/helpers/calls.py [hunk 12]
```diff
@@ -167,6 +167,28 @@ def generate_record():
...
+
+    assert categories
```
Comment (michamos) [inline tests/integration/workflows/helpers/calls.py]: what is this for? you are defining `categories` to be non-empty. If you want to assert that the category is in the record, you should assert that.

File: tests/integration/workflows/test_workflow_actions.py [hunk 13]
```diff
@@ -329,3 +399,53 @@ def test_normalize_journal_titles_unknown_journals_no_ref(workflow_app, insert_j
...
+
+    citing_record, categories = insert_citing_record()
+
+    extra_config = {
+        "BEARD_API_URL": "http://example.com/beard",
+        "MAGPIE_API_URL": "http://example.com/magpie",
+        'ARXIV_CATEGORIES': categories,
+    }
+
+    with workflow_app.app_context():
```
Comment (michamos) [inline tests/integration/workflows/test_workflow_actions.py]: aren't you already in an application context by using the `workflow_app` fixture?

File: tests/unit/workflows/test_workflows_tasks_actions.py [hunk 14]
```diff
@@ -819,77 +818,6 @@ def test_populate_journal_coverage_does_nothing_if_no_journal_is_found(mock_repl
     assert 'journal_coverage' not in obj.extra_data
 
 
-@patch('inspirehep.modules.workflows.tasks.actions.get_document_in_workflow')
```
Comment (michamos) [inline tests/unit/workflows/test_workflows_tasks_actions.py]: why did you remove all these tests?

File: inspirehep/modules/workflows/config.py (hunk 3 above)
Comment (michamos) [reply inspirehep/modules/workflows/config.py]: it's not about reducing the number of queries, but having more matches, when we wrongly identify what is a `page_start` and what is an `artid`.

File: inspirehep/modules/workflows/tasks/refextract.py (hunk 10 above)
Comment (salmanmaq) [reply inspirehep/modules/workflows/tasks/refextract.py]: Nice catch.

File: tests/integration/workflows/test_workflow_actions.py (hunk 13 above)
Comment (salmanmaq) [reply tests/integration/workflows/test_workflow_actions.py]: I am not sure about this. Will check.

File: tests/unit/workflows/test_workflows_tasks_actions.py (hunk 14 above)
Comment (salmanmaq) [reply tests/unit/workflows/test_workflows_tasks_actions.py]: They were unit tests, and they are bound to fail due to inclusion of the inspire matcher since there's no ElasticSearch instance in Unit tests.

File: tests/integration/workflows/helpers/calls.py (hunk 12 above)
Comment (salmanmaq) [reply tests/integration/workflows/helpers/calls.py]: I think this is a very hacky way to get the task done in the first place (to make it auto-accepted). I am  not exactly sure why this was done in the first place in https://github.com/inspirehep/inspire-next/blob/ac4fdbbf50b03b55dd5f4983600ec65cca4879d8/tests/integration/workflows/helpers/calls.py#L170
where I took the inspiration from.

File: tests/integration/workflows/test_workflow_actions.py (hunk 13 above)
Comment (salmanmaq) [reply tests/integration/workflows/test_workflow_actions.py]: It does work without this line, but I am unfamiliar with Flask so I am not completely sure what's going on.

File: tests/integration/workflows/helpers/calls.py (hunk 11 above)
Comment (salmanmaq) [reply tests/integration/workflows/helpers/calls.py]: Just copied from: https://github.com/inspirehep/inspire-next/blob/ac4fdbbf50b03b55dd5f4983600ec65cca4879d8/tests/integration/workflows/helpers/calls.py#L170
Again, not sure why we need to do this.

File: tests/unit/workflows/test_workflows_tasks_actions.py (hunk 14 above)
Comment (michamos) [reply tests/unit/workflows/test_workflows_tasks_actions.py]: then they should become integration tests instead of unit, or mock the reference matching if they don't depend on it.

File: tests/unit/workflows/test_workflows_tasks_actions.py [hunk 15]
```diff
@@ -1124,3 +1125,63 @@ def test_validate_record_raises_when_record_is_invalid():
 
     with pytest.raises(ValidationError):
         _validate_record(obj, eng)
+
+
+@patch('inspirehep.modules.workflows.tasks.actions.get_document_in_workflow')
+@patch(
+    'inspirehep.modules.workflows.tasks.refextract.match',
+    return_value=(_ for _ in [])
```
Comment (michamos) [inline tests/unit/workflows/test_workflows_tasks_actions.py]: why not `iter([])`?

Comment (turtle321) [reply tests/unit/workflows/test_workflows_tasks_actions.py]: My fault, I suggested it. But this is not really important, right? Let's try to merge this today.

Comment (michamos) [reply tests/unit/workflows/test_workflows_tasks_actions.py]: I was scratching my head for a couple of minutes trying to understand this, so I would be in favor of changing it. I agree on trying to get this merged today.

Comment (turtle321) [reply tests/unit/workflows/test_workflows_tasks_actions.py]: Ok

File: tests/integration/workflows/helpers/calls.py [hunk 16]
```diff
@@ -167,6 +167,25 @@ def generate_record():
...
+            '../fixtures',
+            'oai_arxiv_citing_record.json'
+        )
+    ))
+
+    categories = {'core': [], 'non-core': []}
+    for eprint in json_data.get('arxiv_eprints', []):
+        categories['core'].extend(eprint.get('categories', []))
+
+    assert 'core' in categories
```
Comment (michamos) [inline tests/integration/workflows/helpers/calls.py]: You can remove this.

**Comment by salmanmaq** [comment]
ci merge please

**Comment by michamos** [comment]
ci merge please
